### ISSUE TITLE
GHC 8.6.1 build issue:  Dynamic linker not initialised

### ISSUE THREAD

**Issue by deepfire** [goldfirere/th-desugar#96]
I know, not `th-desugar`'s problem but still:
```
[11 of 11] Compiling Language.Haskell.TH.Desugar.Lift ( Language/Haskell/TH/Desugar/Lift.hs, dist/build/Language/Haskell/TH/Desugar/Lift.p_o )
Preprocessing test suite 'spec' for th-desugar-1.8..
Building test suite 'spec' for th-desugar-1.8..
[1 of 4] Compiling Splices          ( Test/Splices.hs, dist/build/spec/spec-tmp/Splices.o )
[2 of 4] Compiling DsDec            ( Test/DsDec.hs, dist/build/spec/spec-tmp/DsDec.o )

<no location info>: error:
    ghc: panic! (the 'impossible' happened)
  (GHC version 8.6.1 for x86_64-unknown-linux):
        Dynamic linker not initialised

Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug

[3 of 4] Compiling Dec              ( Test/Dec.hs, dist/build/spec/spec-tmp/Dec.o )

<no location info>: error:
    <command line>: can't load .so/.DLL for: /run/user/1000/nix-build-th-desugar-1.8.drv-0/source/dist/build/libHSth-desugar-1.8-KFOywxTzF8CA5cwAvr5NgA-ghc8.6.1.so (/run/user/1000/nix-build-th-desugar-1.8.drv-0/source/dist/build/libHSth-desugar-1.8-KFOywxTzF8CA5cwAvr5NgA-ghc8.6.1.so: undefined symbol: thzmdesugarzm1zi8zmKFOywxTzzF8CA5cwAvr5NgA_LanguageziHaskellziTHziDesugarziAST_DAppPr_closure)
builder for '/nix/store/yn634096vg0ih9mk35f2kx3hv80rx71v-th-desugar-1.8.drv' failed with exit code 1
error: build of '/nix/store/yn634096vg0ih9mk35f2kx3hv80rx71v-th-desugar-1.8.drv' failed

[nix-shell:~/holotype]$ cat pkgs/def/github/th-desugar.rev
1d164b669079fe57181af95bbbe4749dbf08ab78
````

**Comment by deepfire** [comment]
I could find this only open GHC bug that _seemed_ relevant: https://ghc.haskell.org/trac/ghc/ticket/14512

**Comment by deepfire** [comment]
https://github.com/goldfirere/singletons/issues/369 might be tangentially related, as it would appear that the interface files generated by `8.6.1` under some circumstances are missing some details.

**Comment by RyanGlScott** [comment]
I'm afraid I can't reproduce this issue locally (on 64-bit Linux). Can you report which `Cabal`/`ghc` bugs you're using to compile `th-desugar`? That might help narrow down what the culprit is here.

**Comment by RyanGlScott** [comment]
And to be clear, I did try building this with Nix as well. This appeared to work without issue on my machine:

```
nix-repl> :b pkgs.haskell.packages.ghc861.callHackage "th-desugar" "1.9" {}
```

So I suspect there's at least one thing that's unique in your setup that I'm not aware of.

**Comment by RyanGlScott** [comment]
Actually, there's an even more looming issue here: I just noticed in your original post that you're trying to build `th-desugar-1.8` with GHC 8.6.1. Why are you doing this? `th-desugar-1.8` doesn't support 8.6.1, as indicated by its restrictive upper version bounds on `template-haskell`.

You really should be using `th-desugar-1.9` if you want to be building with 8.6.1. Can you try building `th-desugar-1.9` and seeing if the issue persists?

**Comment by deepfire** [comment]
@RyanGlScott, the version number is misleading -- it only says what Nix derivation was used before its `src` was overridden.  The real commit ID was pasted in the log -- `1d164b669079fe57181af95bbbe4749dbf08ab78`.

**Comment by RyanGlScott** [comment]
Hm, OK. Again, I'm not super-familiar with what Nix is doing under the hood, so having a build log that shows exactly what flags are being used would be incredibly helpful for me to debug this issue.

**Comment by deepfire** [comment]
@RyanGlScott, will do a bit later!

**Comment by deepfire** [comment]
Here is the output of the configuration phase:
```
Using Parsec parser
Configuring th-desugar-1.8...
Dependency base -any: using base-4.12.0.0
Dependency containers -any: using containers-0.6.0.1
Dependency mtl -any: using mtl-2.2.2
Dependency syb -any: using syb-0.7
Dependency template-haskell -any: using template-haskell-2.14.0.0
Dependency th-expand-syns -any: using th-expand-syns-0.4.4.0
Dependency th-lift -any: using th-lift-0.7.11
Dependency th-orphans -any: using th-orphans-0.13.6
Source component graph: component lib
Configured component graph:
    component th-desugar-1.8-GogzLGV4txJHGzPgbKc3X
        include base-4.12.0.0
        include containers-0.6.0.1
        include mtl-2.2.2
        include syb-0.7-8OHoa6XfXrB5PPoB0H7POz
        include template-haskell-2.14.0.0
        include th-expand-syns-0.4.4.0-JJJC0BlUtveGlxXaCaUH0k
        include th-lift-0.7.11-JiIK3cdnl9dHd8tTTTgRnz
        include th-orphans-0.13.6-B1Syz1RY8XYYSOIt3RHv5
Linked component graph:
    unit th-desugar-1.8-GogzLGV4txJHGzPgbKc3X
        include base-4.12.0.0
        include containers-0.6.0.1
        include mtl-2.2.2
        include syb-0.7-8OHoa6XfXrB5PPoB0H7POz
        include template-haskell-2.14.0.0
        include th-expand-syns-0.4.4.0-JJJC0BlUtveGlxXaCaUH0k
        include th-lift-0.7.11-JiIK3cdnl9dHd8tTTTgRnz
        include th-orphans-0.13.6-B1Syz1RY8XYYSOIt3RHv5
        Language.Haskell.TH.Desugar=th-desugar-1.8-GogzLGV4txJHGzPgbKc3X:Language.Haskell.TH.Desugar,Language.Haskell.TH.Desugar.Expand=th-desugar-1.8-GogzLGV4txJHGzPgbKc3X:Language.Haskell.TH.Desugar.Expand,Language.Haskell.TH.Desugar.Lift=th-desugar-1.8-GogzLGV4txJHGzPgbKc3X:Language.Haskell.TH.Desugar.Lift,Language.Haskell.TH.Desugar.Subst=th-desugar-1.8-GogzLGV4txJHGzPgbKc3X:Language.Haskell.TH.Desugar.Subst,Language.Haskell.TH.Desugar.Sweeten=th-desugar-1.8-GogzLGV4txJHGzPgbKc3X:Language.Haskell.TH.Desugar.Sweeten
Ready component graph:
    definite th-desugar-1.8-GogzLGV4txJHGzPgbKc3X
        depends base-4.12.0.0
        depends containers-0.6.0.1
        depends mtl-2.2.2
        depends syb-0.7-8OHoa6XfXrB5PPoB0H7POz
        depends template-haskell-2.14.0.0
        depends th-expand-syns-0.4.4.0-JJJC0BlUtveGlxXaCaUH0k
        depends th-lift-0.7.11-JiIK3cdnl9dHd8tTTTgRnz
        depends th-orphans-0.13.6-B1Syz1RY8XYYSOIt3RHv5
Using Cabal-2.4.0.1 compiled by ghc-8.6
Using compiler: ghc-8.6.1
Using install prefix:
/nix/store/9srgihvizm86r9r5kynxr5d6qdv4fs5d-th-desugar-1.8
Executables installed in:
/nix/store/9srgihvizm86r9r5kynxr5d6qdv4fs5d-th-desugar-1.8/bin
Libraries installed in:
/nix/store/9srgihvizm86r9r5kynxr5d6qdv4fs5d-th-desugar-1.8/lib/ghc-8.6.1/x86_64-linux-ghc-8.6.1/th-desugar-1.8-GogzLGV4txJHGzPgbKc3X
Dynamic Libraries installed in:
/nix/store/9srgihvizm86r9r5kynxr5d6qdv4fs5d-th-desugar-1.8/lib/ghc-8.6.1/x86_64-linux-ghc-8.6.1
Private executables installed in:
/nix/store/9srgihvizm86r9r5kynxr5d6qdv4fs5d-th-desugar-1.8/libexec/x86_64-linux-ghc-8.6.1/th-desugar-1.8
Data files installed in:
/nix/store/9srgihvizm86r9r5kynxr5d6qdv4fs5d-th-desugar-1.8/share/x86_64-linux-ghc-8.6.1/th-desugar-1.8
Documentation installed in:
/nix/store/sc3r4c1kmkqr4hg0by1ap1z7gs97qbz6-th-desugar-1.8-doc/share/doc/th-desugar-1.8
Configuration files installed in:
/nix/store/9srgihvizm86r9r5kynxr5d6qdv4fs5d-th-desugar-1.8/etc
No alex found
Using ar found on system at:
/nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/ar
No c2hs found
No cpphs found
No doctest found
Using gcc version 7.3.0 given by user at:
/nix/store/10yq7kwlvbc6h658izmrlsspry1g9f3c-gcc-wrapper-7.3.0/bin/gcc
Using ghc version 8.6.1 found on system at:
/nix/store/qlh14liv4d8hkkkj7m5i9nyx04yv87hc-ghc-8.6.1/bin/ghc
Using ghc-pkg version 8.6.1 found on system at:
/nix/store/qlh14liv4d8hkkkj7m5i9nyx04yv87hc-ghc-8.6.1/bin/ghc-pkg
No ghcjs found
No ghcjs-pkg found
No greencard found
Using haddock version 2.20.0 found on system at:
/nix/store/qlh14liv4d8hkkkj7m5i9nyx04yv87hc-ghc-8.6.1/bin/haddock
No happy found
Using haskell-suite found on system at: haskell-suite-dummy-location
Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
No hmake found
Using hpc version 0.67 found on system at:
/nix/store/qlh14liv4d8hkkkj7m5i9nyx04yv87hc-ghc-8.6.1/bin/hpc
Using hsc2hs version 0.68.4 found on system at:
/nix/store/qlh14liv4d8hkkkj7m5i9nyx04yv87hc-ghc-8.6.1/bin/hsc2hs
Using hscolour version 1.24 found on system at:
/nix/store/lzg6ws0jqk71p0g6aqlja6win16i0057-hscolour-1.24.4/bin/HsColour
No jhc found
Using ld found on system at:
/nix/store/an3icm1hajqx1lld6n8q4qr40nwbs544-binutils-wrapper-2.30/bin/ld
No pkg-config found
Using runghc version 8.6.1 found on system at:
/nix/store/qlh14liv4d8hkkkj7m5i9nyx04yv87hc-ghc-8.6.1/bin/runghc
Using strip version 2.30 found on system at:
/nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
Using tar found on system at:
/nix/store/y9nwn094yhwkg11pdl1w3427lyk9nvg3-gnutar-1.30/bin/tar
No uhc found
```

**Comment by RyanGlScott** [comment]
Unfortunately, this doesn't really tell me much besides the installed dependencies (which are more-or-less identical on my machine).

Do you have access to the invocation of `runghc Setup.hs build` (IIRC, that's what Nix does behind the scenes)? Preferably, that would include the list of `Cabal` flags passed to it.

**Comment by deepfire** [comment]
I think I'm closing in on the issue, there appears to be a couple of lines in the cabal file that trigger the error. A bit more time..

**Comment by RyanGlScott** [comment]
Thanks.

Before I spend time digging deeper into this, I feel obligated to ask this question: is there a reason why you're removing things from `other-modules`? That's almost guaranteed to cause problems of various sorts—a notable example of an issue that this has caused in the past includes guillaume-nargeot/hpc-coveralls#52 (when a package was compiled with `--enable-executable-dynamic`). It's also why the [`-Wmissing-home-modules` warning](https://ghc.haskell.org/trac/ghc/ticket/13129) was introduced in the first place to warn about this sort of thing.

My hunch is that if you added these modules back, this issue (and goldfirere/singletons#369) would go away.

**Comment by deepfire** [comment]
Ryan, absolutely, those two lines make the entire difference -- they are the reproduction for me.

And indeed, that's the wrong way to use Cabal -- but the larger question is should GHC panic if you have a malformed Cabal file?

**Comment by RyanGlScott** [comment]
OK, I just wanted to make it clear that this setup shouldn't be expected to work.

I was able to reproduce the panic with another change on top of deepfire@c5044fa:

```diff
diff --git a/cabal.project b/cabal.project
index e6fdbad..89344a8 100644
--- a/cabal.project
+++ b/cabal.project
@@ -1 +1,4 @@
 packages: .
+
+package th-desugar
+  ghc-options: -j2
```

With this, you can `cabal new-build --enable-tests` and get the panic for yourself:

```
$ cabal new-build --enable-tests
Build profile: -w ghc-8.6.1 -O1
In order, the following will be built (use -v for more details):
 - th-desugar-1.9 (test:spec) (first run)
Preprocessing test suite 'spec' for th-desugar-1.9..
Building test suite 'spec' for th-desugar-1.9..
[2 of 4] Compiling DsDec            ( Test/DsDec.hs, /home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.6.1/th-desugar-1.9/t/spec/build/spec/spec-tmp/DsDec.o )

<no location info>: error:
    <command line>: can't load .so/.DLL for: /home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.6.1/th-desugar-1.9/build/libHSth-desugar-1.9-inplace-ghc8.6.1.so (/home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.6.1/th-desugar-1.9/build/libHSth-desugar-1.9-inplace-ghc8.6.1.so: undefined symbol: thzmdesugarzm1zi9zminplace_LanguageziHaskellziTHziDesugarziAST_DWildPa_closure)
[3 of 4] Compiling Dec              ( Test/Dec.hs, /home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.6.1/th-desugar-1.9/t/spec/build/spec/spec-tmp/Dec.o )

<no location info>: error:
    ghc: panic! (the 'impossible' happened)
  (GHC version 8.6.1 for x86_64-unknown-linux):
        Dynamic linker not initialised

Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug
```

As far as the panic itself goes, I now remember that this is actually an old GHC bug: [Trac #13531](https://ghc.haskell.org/trac/ghc/ticket/13531). This has been fixed in GHC HEAD (and will debut in GHC 8.8.1). I just confirmed that GHC HEAD doesn't panic with the same setup:

```
$ cabal new-build --enable-tests -w /opt/ghc/head/bin/ghc --keep-going
Build profile: -w ghc-8.7.20181015 -O1
In order, the following will be built (use -v for more details):
 - th-desugar-1.9 (test:spec) (first run)
Preprocessing test suite 'spec' for th-desugar-1.9..
Building test suite 'spec' for th-desugar-1.9..
[2 of 4] Compiling DsDec            ( Test/DsDec.hs, /home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.7.20181015/th-desugar-1.9/t/spec/build/spec/spec-tmp/DsDec.o )

<no location info>: error:
    <command line>: can't load .so/.DLL for: /home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.7.20181015/th-desugar-1.9/build/libHSth-desugar-1.9-inplace-ghc8.7.20181015.so (/home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.7.20181015/th-desugar-1.9/build/libHSth-desugar-1.9-inplace-ghc8.7.20181015.so: undefined symbol: thzmdesugarzm1zi9zminplace_LanguageziHaskellziTHziDesugarziAST_DWildPa_closure)
[3 of 4] Compiling Dec              ( Test/Dec.hs, /home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.7.20181015/th-desugar-1.9/t/spec/build/spec/spec-tmp/Dec.o )

<no location info>: error:
    <command line>: can't load .so/.DLL for: /home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.7.20181015/th-desugar-1.9/build/libHSth-desugar-1.9-inplace-ghc8.7.20181015.so (/home/rgscott/Documents/Hacking/Haskell/th-desugar/dist-newstyle/build/x86_64-linux/ghc-8.7.20181015/th-desugar-1.9/build/libHSth-desugar-1.9-inplace-ghc8.7.20181015.so: undefined symbol: thzmdesugarzm1zi9zminplace_LanguageziHaskellziTHziDesugarziAST_DWildPa_closure)
```

I'm opting to close this issue, since it falls squarely into GHC-bug territory (and not `th-desugar`-bug territory). Please re-open if you disagree.

**Comment by deepfire** [comment]
Excellent news, @RyanGlScott, thank you a lot!
